File: C++/Lista/fila.cpp
#include "fila.h"

Fila::Fila() : Lista() {}

Fila::Fila(int d) : Lista(d) {}

Fila::Fila(const Fila& f) : Lista(f) {}

Fila::~Fila() {}

int Fila::remove() {
    if (minhaLista == nullptr) {
        return -1; // Fila vazia
    }

    No* temp = minhaLista;
    No* anterior = nullptr;

    while (temp->get() != nullptr) {
        anterior = temp;
        temp = temp->get();
    }

    int valorRemovido = temp->getDado();

    if (anterior != nullptr) {
        delete temp;
        anterior->prox = nullptr;
    } else {
        // Removendo o único elemento da fila
        delete temp;
        minhaLista = nullptr;
    }

    tamLista--;

    return valorRemovido;
}

int Fila::primeiro() {
    if (minhaLista == nullptr) {
        return -1; // Fila vazia
    }

    No* temp = minhaLista;

    while (temp->get() != nullptr) {
        temp = temp->get();
    }

    return temp->getDado();
}

File: PrimeiraParte/Java/Lista/Fila.java
public class Fila extends Lista {
    public Fila() {
        super();
    }

    public Fila(int d) {
        super(d);
    }

    // Métodos específicos da fila
    public void enfileira(int d) {
        insere(d);
    }

    public int desenfileira() {
        if (minhaLista == null) {
            return -1;  // Fila vazia
        }
        No temp = minhaLista;
        No tempAnterior = null;
        while (temp.prox != null) {
            tempAnterior = temp;
            temp = temp.prox;
        }
        int valorRemovido = temp.dado;
        if (tempAnterior == null) {
            minhaLista = null;
        } else {
            tempAnterior.prox = null;
        }
        tamLista--;
        return valorRemovido;
    }

    public int frente() {
        if (minhaLista == null) {
            return -1;  // Fila vazia
        }
        return minhaLista.dado;
    }
}

